src/boot/chid.h

```diff
@@ -45,20 +55,28 @@ assert_cc(offsetof(Device, descriptor) == 0);
 assert_cc(offsetof(Device, chid) == 4);
 assert_cc(offsetof(Device, devicetree.name_offset) == 20);
 assert_cc(offsetof(Device, devicetree.compatible_offset) == 24);
+assert_cc(offsetof(Device, fw.name_offset) == 20);
+assert_cc(offsetof(Device, fw.compatible_offset) == 24);
 assert_cc(sizeof(Device) == 28);
 
-static inline const char* device_get_name(const void *base, const Device *device) {
-        if (device->descriptor != DEVICE_DESCRIPTOR_DEVICETREE)
+static inline const char* device_get_name(const void *base, const Device *device, uint32_t descriptor) {
+        if (descriptor == DEVICE_DESCRIPTOR_DEVICETREE) {
```

Comment:
I guess you meant `device->descriptor` here and in all the other places. This should be a switch-case for `DEVICE_TYPE_FROM_DESCRIPTOR(device->descriptor)`.

Maybe also just copy `{name,compatible}_offset` to a local variable and then return `off == 0 ? NULL : (const char *) ((const uint8_t *) base + off))` in the end of the function. It's quite easy to make copy-paste error when adding a new type here.

Reply:
> I guess you meant `device->descriptor` here and in all the other places. This should be a switch-case for `DEVICE_TYPE_FROM_DESCRIPTOR(device->descriptor)`.

What really do you want? I am so confused and this type of back and forth is so frustrating! Can you write in code the exact changes you want to see?
I did not mean `device->descriptor` there because based upon the type of device, it has to access the right structures. Clearly, that is not what you would like to see.

Reply:
> Caller of chid_match (be it pe_locate_sections or wherever this FW thing would be handled) always expects a very specific kind of device to be returned, so chid_match should probably take the type as a parameter and ignore entries with different types in the matching process.
> @poettering what do you think about this?

What did you really mean by this in the other thread? Does `device_get_name` etc take a third parameter as descriptor or not?

Reply:
> I did not mean `device->descriptor` there because based upon the type of device, it has to access the right structures. Clearly, that is not what you would like to see.

Because it's wrong and inconsistent. `device->descriptor` is the field that contains the type. `device_get_*` functions are just simple helpers to get the strings, they are only needed to not write that ternary operator over and over again. These functions do not need another `descriptor` parameter because of that.

Reply:
> What did you really mean by this in the other thread? Does `device_get_name` etc take a third parameter as descriptor or not?

`device_get_name` and `chid_match` are the different functions, so not sure what's not clear here. The latter needs another parameter (and it should be type, not descriptor btw), the former doesn't.